### PR TITLE
Implicit tiling

### PR DESCRIPTION
**Pull request by loshjawrence**
DRAFT @pjcozzi

### CODE REVIEW

**Comment by pjcozzi** [comment]
Can we retarget this pull request to a new staging branch something like `3d-tiles-new-extensions`?  We can queue up the upcoming extensions there then merge in `master` when everything is in good shape.

File: extensions/3DTILES_implicit_tiling_scheme/README.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+oct subtree: [0b00000001, 0b11111111, 0b11111111, ...]
+
+Bits are left to right, top to bottom raster order. LSB bits are earlier in the raster order.
+
+Note: Padding bits in the root of the subtree can allow the subtree itself communicate how deep it goes. Will let implementation dictate that this is more desirable than fix sizes (don't think it will, hasn't yet).
+
+Below is a binary subtree of 4 levels. There are two leaf tiles at level 3 (root is level 0) that are available.
+These would have tiles available for requesting (like all the other 1's) but they would also have subtree binaries available for requesting as well, at uri "baseUri/availability/d/x" (d/x in this case since its binary. quad would be d/x/y, oct would be d/x/y/z)
+
+![](subtreeBits.jpg)
```
[the remark this answers is not in the document]
Comment (loshjawrence) [reply extensions/3DTILES_implicit_tiling_scheme/README.md]: My bad, I must have misheard. Please recap, in mean time I can go back and draw something simple for each attribute and how it relates to the physical tree.

[the remark this answers is not in the document]
Comment (pjcozzi) [reply extensions/3DTILES_implicit_tiling_scheme/README.md]: I _think_ I ended up requesting all the figures in other parts of this PR.  Let's start with those and then we'll see if anything else would be useful.

**Comment by loshjawrence** [comment]
@shehzan10 Please review when you get a chance. I did not cover all of Patrick's comments. This is taking long than I expected and need to switch to implementation since I'm getting behind on that.

**Comment by pjcozzi** [comment]
@loshjawrence one more question: we discussed that availability should be more general purpose than just boolean for its presence, bounding volume, number of triangles, etc.  However, we don't want the overhead of a JSON object per tile for this metadata so we discussed the ability to define a general schema and then each tile to store a small binary blob in that schema.  Is this something covered by this extension?  Or is that separate?

**Comment by loshjawrence** [comment]
Separate, it lives here atm https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/383

**Comment by loshjawrence** [comment]
@shehzan10 @lilleyse Can you guys add what ever comments you may have when you get a chance?

**Comment by pjcozzi** [comment]
> @shehzan10 @lilleyse Can you guys add what ever comments you may have when you get a chance?

Bump @shehzan10 @lilleyse

**Comment by pjcozzi** [comment]
@loshjawrence one more question - ContextCapture can export tiled OBJs - how does that tiling map to 3D Tiles implicit tiling?  And can you please include an example in the spec?

**Comment by loshjawrence** [comment]
It sounds like we currently take the tiled obj leaves and tile those through the beast. When I talked to Gary last about getting the beast to output implicit tilesets he it would be matter of swapping out a function. @likangning93 feel free to correct me if I'm mis-quoting here.

File: extensions/3DTILES_implicit_tiling_scheme/README.md [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+The full tree of information that expresses all tiles' availability is broken up into subtrees(small portions of the full tree). Since a single bit is needed to hold a tile's availability,
+the subtree of availability is expressed as an array of bytes where each bit holds a tile's availability, i.e. a 1 or 0 indicating that the tile is available or not available, respectively.
+
+### Complete Subtree Levels
+In the `tileset.json`, the `completeSubtreeLevels` property is the number of levels in every subtree.
+Every subtree starts from a single tile, its root, and goes down the number levels indicated by `completeSubtreeLevels`. For example,
+```json
+`completeSubtreeLevels: 10`
+```
+means that if a subtree started at level 0, it would go down to level 9, covering 10 levels.
```
Comment (lilleyse) [inline extensions/3DTILES_implicit_tiling_scheme/README.md]: It should be more clearly stated that subtrees are aligned to `completeSubtreeLevels`. If `completeSubtreeLevels` is 9 `firstSubtreesWithContent` must start at either level 0, 8, 16, etc. Also provide some context for this choice - that it makes it easier to fuse tilesets or do partial updates.

**Comment by lilleyse** [comment]
Can we encode more metadata in the subtree and go beyond just availability?

A generic / extensible JSON object in header would open up more possibilities for storing metadata. E.g. tighter bounding volume, subtree stats (number of tiles available), subtree depth if it's less than `completeSubtreeLevels`.Even availability could be a JSON property with a byte offset / byte length into some binary section, modeled after the glTF / batch table/ feature paradigm.

This idea will help make the subtree/availability concept more extensible and future-proof.

**Comment by lilleyse** [comment]
If implicit tiling is going to be used for vector and raster data we may need to think more about Web Mercator. It would be nice if a tileset.json with the implicit tiling extension could be compatible with existing imagery tilesets without the need to unproject to WGS84. At the very least the tileset.json would need to constrain its region to `-180.0 -85.06 180.0 85.06` and state the projection somewhere.

@tfili put this list together of various imagery tiling schemes in another issue:

> TMS - WGS84 1x2 or Mercator 1x1 depending on xml file. Will default to mercator.
WMS - Defaults to WGS84 1x2 if not specified on provider creation.
WMTS - Defaults to Mercator 1x1 if not specified on provider creation.
Bing - Mercator 2x2
Google Earth - WGS84 2x2

**Comment by lilleyse** [comment]
Related but not necessarily part of this extension - tiles should have a metadata store too, somewhere to put their bounding volume, content bounding volume, children bounding volumes (potentially) for optimal culling at runtime. Quantized-mesh stores does this too.

We first started thinking about this for tile expiration (https://github.com/AnalyticalGraphicsInc/3d-tiles/pull/217) where new tiles might have different bounding volumes or expiration than what was in tileset.json.

**Comment by loshjawrence** [comment]
For region tilesets, this needs to add something like:
```
globalContext: {
    boundingVolume: {
        region:[...]
    }
}
```
Local tileset should use this too, i.e. some subset is checkouted updates are added and needs to be integrated back to the source with correct alignment, keys, subtrees.

**Comment by sanjeetsuhag** [comment]
Some notes from offline discussion with @lilleyse 

- How do we want to handle the file path with external tilesets? Using an approach similar to symlinks could work.
- How do we want to store the feature table data at the subtree level? Storing the `name` property of `EXT_3dtiles_feature_metadata` could work.
- How do we handle the inaccuracy that S2 regions (using spherical geodesics) introduces?
- How do we keep the availability object size minimal? We want to keep availability as small as possible, to store information vital for rendering and traversal, but metadata should be kept separate. Need to investigate if this breaks random access when looking up metadata.

Some questions for VRICON:

1. Will different subtrees store different metadata within a tileset?
2. Will different tiles within a subtree store different metadata?

**Comment by lilleyse** [comment]
A second draft of this extension is here: https://github.com/CesiumGS/3d-tiles/pull/414

**Pull request closed without merging**